File: 0227-basic-calculator-ii/0227-basic-calculator-ii.py
class Solution:
    def calculate(self, s: str) -> int:
        prev_op = '+'
        num = 0 
        stck = []

        for idx, c in enumerate(s):
            if c.isdigit():
                num = num * 10 + int(c)
            if not c.isdigit() and c != ' ' or idx == len(s) - 1:
                if prev_op == '+':
                    stck.append(num)
                elif prev_op == '-':
                    stck.append(-num)
                elif prev_op == '*':
                    stck.append(stck.pop() * num)
                elif prev_op == '/':
                    stck.append(int(stck.pop() / num))
                prev_op = c
                num = 0
        return sum(stck)

File: 0227-basic-calculator-ii/0227-basic-calculator-ii.cpp
class Solution {
public:
    int calculate(string s) {
        stack<int> stck;
        char prev_op = '+';
        int num = 0;

        for (int i = 0; i < s.size(); i++) {
            char c = s[i];

            if (isdigit(c)) {
                num = num * 10 + (c - '0');
            }

            if (!isdigit(c) && c != ' ' || i == s.size() - 1) {
                if (prev_op == '+') {
                    stck.push(num);
                }
                else if (prev_op == '-') {
                    stck.push(-num);
                }
                else if (prev_op == '*') {
                    int top = stck.top();
                    stck.pop();
                    stck.push(top * num);
                }
                else if (prev_op == '/') {
                    int top = stck.top();
                    stck.pop();
                    stck.push(top / num);
                }

                prev_op = c;
                num = 0;
            }
        }
        
        int result = 0;

        while(!stck.empty()) {
            int top = stck.top();
            result += top;
            stck.pop();
        }

        return result;
    }
};